base/linalg/generic.jl

```diff
@@ -1165,8 +1165,46 @@ function logdet(A::AbstractMatrix)
     return d + log(s)
 end
 
+"""
+    promote_leaf_eltypes(itr)
+
+For an (possibly nested) iterable object `itr`, promote the types of leaf
+elements.  Equivalent to `promote_type(typeof(leaf1), typeof(leaf2), ...)`.
+
+# Example
+
+```jldoctest
+julia> a = [[1,2, [3,4]], 5.0, [6im, [7.0, 8.0]]]
+3-element Array{Any,1}:
+  Any[1,2,[3,4]]
+ 5.0
+  Any[0+6im,[7.0,8.0]]
+
+julia> promote_leaf_eltypes(a)
+Complex{Float64}
+```
+"""
+function promote_leaf_eltypes(itr)
+    if !isa(itr, AbstractArray) && !isa(itr, Tuple)  # handles arrays, tuples, ranges
+        return typeof(itr)
```

Comment:
The normal Julian thing is to use dispatch rather than explicit typechecks:
```jl
promote_leaf_eltypes(itr::Union{AbstractArray,Tuple}) = ... do something ...
promote_leaf_eltypes{T}(itr{T}) = T
```